typescript/sdk/src/token/Token.ts

```diff
@@ -0,0 +1,343 @@
+/* eslint-disable @typescript-eslint/no-empty-interface */
+import {
+  Address,
+  Numberish,
+  ProtocolType,
+  eqAddress,
+} from '@hyperlane-xyz/utils';
+
+import { ChainMetadata } from '../metadata/chainMetadataTypes';
+import { MultiProtocolProvider } from '../providers/MultiProtocolProvider';
+import { ChainName } from '../types';
+
+import { TokenAmount } from './TokenAmount';
+import {
+  PROTOCOL_TO_NATIVE_STANDARD,
+  TOKEN_HYP_STANDARDS,
+  TOKEN_MULTI_CHAIN_STANDARDS,
+  TOKEN_NFT_STANDARDS,
+  TOKEN_STANDARD_TO_PROTOCOL,
+  TokenStandard,
+} from './TokenStandard';
+import {
+  CwHypCollateralAdapter,
+  CwHypNativeAdapter,
+  CwHypSyntheticAdapter,
+  CwNativeTokenAdapter,
+  CwTokenAdapter,
+} from './adapters/CosmWasmTokenAdapter';
+import {
+  CosmIbcTokenAdapter,
+  CosmNativeTokenAdapter,
+} from './adapters/CosmosTokenAdapter';
+import {
+  EvmHypCollateralAdapter,
+  EvmHypNativeAdapter,
+  EvmHypSyntheticAdapter,
+  EvmNativeTokenAdapter,
+  EvmTokenAdapter,
+} from './adapters/EvmTokenAdapter';
+import { IHypTokenAdapter, ITokenAdapter } from './adapters/ITokenAdapter';
+import {
+  SealevelHypCollateralAdapter,
+  SealevelHypNativeAdapter,
+  SealevelHypSyntheticAdapter,
+  SealevelNativeTokenAdapter,
+  SealevelTokenAdapter,
+} from './adapters/SealevelTokenAdapter';
+
+export interface TokenArgs {
+  chainName: ChainName;
+  standard: TokenStandard;
+  decimals: number;
+  symbol: string;
+  name: string;
+  addressOrDenom: Address | string;
+  collateralAddressOrDenom?: Address | string;
+  igpTokenAddressOrDenom?: string;
+  logoURI?: string;
+  connectedTokens?: Token[];
+
+  // Cosmos specific:
+  sourcePort?: string;
+  sourceChannel?: string;
+}
+
+// Declaring the interface in addition to class allows
+// Typescript to infer the members vars from TokenArgs
+export interface Token extends TokenArgs {}
+
+export class Token {
+  public readonly protocol: ProtocolType;
+
+  constructor(args: TokenArgs) {
+    Object.assign(this, args);
+    this.protocol = TOKEN_STANDARD_TO_PROTOCOL[this.standard];
+  }
+
+  static FromChainMetadataNativeToken(chainMetadata: ChainMetadata): Token {
+    if (!chainMetadata.nativeToken)
+      throw new Error(
+        `ChainMetadata for ${chainMetadata.name} missing nativeToken`,
+      );
+
+    const { protocol, name: chainName, nativeToken, logoURI } = chainMetadata;
+    return new Token({
+      chainName,
+      standard: PROTOCOL_TO_NATIVE_STANDARD[protocol],
+      addressOrDenom: nativeToken.denom ?? '',
+      decimals: nativeToken.decimals,
+      symbol: nativeToken.symbol,
+      name: nativeToken.name,
+      logoURI,
+    });
+  }
+
+  /**
+   * Returns a TokenAdapter for the token and multiProvider
+   * @throws If multiProvider does not contain this token's chain.
+   * @throws If token is an NFT (TODO NFT Adapter support)
+   */
+  getAdapter(multiProvider: MultiProtocolProvider): ITokenAdapter {
```

Comment:
Yes, I realize this method and getHypAdapter are pretty much just maps (type to adapter) but the subtle differences in the adapters and their constructor args make it necessary to spell this out explicitly.